File: Examples/TS/test9.ts
import { Callable } from "./collections.abc";
import { dataclass } from "./dataclasses";
type Tree<A, B> = Leaf<A, B> | LNode<A, B>;
interface Leaf<A, B> {
    value: A;
}
interface LNode<A, B> {
    value: B;
    left: Tree<A, B>;
    right: Tree<A, B>;
}
function max(x: number, y: number): number {
    if (x > y) {
        return x;
    } else {
        return y;
    }
}
function first<A, B>(z: [A, B]): A {
    let x=z, _=z;
    return x;
}
function size(t: Tree<number, string>): number {
    if (t.kind === "Leaf") {
        const value: _ = t.value;
        
        return 1;
    } else if (t.kind === "LNode") {
        const value: _, left: l, right: r = t.value;
        
        return (1 + max(size(l), size(r)));
    }
}
function simple(t: Tree<number, string>): boolean {
    if (t.kind === "Leaf") {
        const value: _ = t.value;
        
        return true;
    } else if (t.kind === "_") {
        const  = t.value;
        
        return false;
    }
}
function compose<A, B, C>(f: (arg0: A) => B, g: (arg0: B) => C): (arg0: A) => C {
    return ((x) => g(f(x)));
}
function succ(x: number): number {
    return (x + 1);
}
function string_of_int(x: number): string {
    return str(x);
}
const f = compose(succ, string_of_int)
const l: Tree<number, string> = { kind: "Leaf", value: { value: 2 } }
const t: Tree<number, string> = { kind: "LNode", value: { value: "ok", left: { kind: "Leaf", value: { value: 3 } }, right: { kind: "Leaf", value: { value: 5 } } } }

print(size(t))

File: Examples/PY/test9.py
from collections.abc import Callable
from dataclasses import dataclass
type Tree[A, B] = Leaf[A, B] | LNode[A, B]
@dataclass
class Leaf[A, B]:
    value: A
@dataclass
class LNode[A, B]:
    value: B
    left: Tree[A, B]
    right: Tree[A, B]
def max(x: int, y: int) -> int:
    if x > y:
        return x
    else:
        return y
def first[A, B](z: tuple[A, B]) -> A:
    (x, _) = z
    return x
def size(t: Tree[int, str]) -> int:
    match t:
        case Leaf(value=_):
            return 1
        case LNode(value=_, left=l, right=r):
            return 1 + (max(size(l), size(r)))
def simple(t: Tree[int, str]) -> bool:
    match t:
        case Leaf(value=_):
            return True
        case _:
            return False
def compose[A, B, C](f: Callable[[A], B], g: Callable[[B], C]) -> Callable[[A], C]:
    return lambda x: g(f(x))
def succ(x: int) -> int:
    return x + 1
def string_of_int(x: int) -> str:
    return str(x)
f = compose(succ, string_of_int)
l: Tree[int, str] = Leaf(value=2)
t: Tree[int, str] = LNode(value="ok", left=Leaf(value=3), right=Leaf(value=5))

print(size(t))
